x/contracts/ethwrapper/client.go

```diff
@@ -0,0 +1,236 @@
+	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
+	NonceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (uint64, error)
+	FilterLogs(ctx context.Context, query ethereum.FilterQuery) ([]types.Log, error)
+}
+
+// errRetry is returned when retry maxRetries is exhausted.
+var errRetry = errors.New("retry attempts exhausted")
+
+// EthClientOptions is a functional option for Client.
+type EthClientOptions func(*Client)
+
+// EthClientWithBlockNumberDrift sets the block number drift
+// for the BlockNumber method.
+func EthClientWithBlockNumberDrift(drift int) EthClientOptions {
+	return func(c *Client) { c.blockNumberDrift = drift }
+}
+
+// EthClientWithWinnersOverride randomly sets the winner in
+// the block header extra data for the HeaderByNumber method.
+func EthClientWithWinnersOverride(winners []string) EthClientOptions {
+	return func(c *Client) { c.winnersOverride = winners }
+}
+
+// EthClientWithMaxRetries sets the maximum number
+// of retries on error for all rpc connections.
+func EthClientWithMaxRetries(retries int) EthClientOptions {
+	return func(c *Client) { c.maxRetries = retries }
+}
+
+// NewClient creates a new ethclient with the given RPC URLs.
+func NewClient(logger *slog.Logger, rpcUrls []string, opts ...EthClientOptions) (*Client, error) {
+	c := &Client{logger: logger}
+
+	var errs error
+	for _, url := range rpcUrls {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		cli, err := ethclient.DialContext(ctx, url)
+		cancel()
+		if err != nil {
+			errs = errors.Join(errs, fmt.Errorf("failed to dial client RPC URL %s: %w", url, err))
+			continue
+		}
+		c.clients = append(
+			c.clients,
+			struct {
+				url string
+				cli EthClient
+			}{
+				url,
+				cli,
+			},
+		)
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+	return c, nil
+}
+
+// Client is an Ethereum client that can connect to multiple RPCs.
+// If an operation fails, it will automatically try the next client node in the list.
+// When all client nodes are exhausted, it will retry the operation up to maxRetries times.
+type Client struct {
+	logger  *slog.Logger
+	clients []struct {
+		url string
+		cli EthClient
+	}
+
+	// Options.
+	blockNumberDrift int
+	winnersOverride  []string
+	maxRetries       int
+}
+
+// RawClient returns the first raw ethclient.
+func (c *Client) RawClient() *ethclient.Client {
+	client, ok := c.clients[0].cli.(*ethclient.Client)
+	if !ok {
+		return nil
+	}
+	return client
+}
+
+// BlockNumber returns the latest block number.
+func (c *Client) BlockNumber(ctx context.Context) (uint64, error) {
+	var errs error
+	for i := range c.maxRetries {
+		for _, client := range c.clients {
+			switch bn, err := client.cli.BlockNumber(ctx); {
+			case err == nil:
+				c.logger.Debug("get block number succeeded", "attempt", i, "block", bn)
+				return bn - uint64(c.blockNumberDrift), nil
```

Comment:
Simply subtracting 64 blocks from the latest block number will work (since two epochs is max finalization time), but sometimes the block will actually be finalized sooner. This could be observed using the "finalized" option https://ethereum.org/en/developers/docs/apis/json-rpc/, but this may be too big of a refactor, as I believe the geth client doesn't support this finalized option by default

Reply:
I have managed to add this to our ethclient wrapper. Can you check again?... Also just doing it for L1